### ISSUE TITLE
Singleton scope not honored when binding to subclasses

### ISSUE THREAD

**Issue by Nyshiki** [alecthomas/injector#99]
I surely got something wrong but I've encountered this unexpected behavior.
When I bind a base class to a singleton subclass I expected just one instance to be created for both dependencies, two different instances were created instead.

package version == 0.14.1

Here's how to reproduce the issue.
```
from abc import ABCMeta, abstractmethod
from injector import singleton, Module, Injector

class IReader(metaclass=ABCMeta):
    @abstractmethod
    def read(self):
        pass

@singleton
class Catalog(IReader):
    def read(self):
        pass

    def add_item(self, item: str):
        pass

class CatalogModule(Module):
    def configure(self, binder):
        binder.bind(IReader, Catalog)

injector = Injector([CatalogModule])

injector.get(IReader) is injector.get(Catalog)
False

injector.get(Catalog)
<Catalog object at 0x7f9d34c83278>

injector.get(IReader) 
<Catalog object at 0x7f9d35786fd0>

```

Hope it is clear enough.
Is this the expected outcome?
Please let me know how to obtain the same instance for both dependencies.

Thanks

**Comment by juharris** [comment]
This does seem like a bug. There could be some reason for it since you might not want this behavior in all cases.

If you really need to get around this:
```python
from injector import provider
```
then in `CatalogModule`:
```python
    @provider
    @singleton
    def provide_catalog(self, reader: IReader) -> Catalog:
        return reader
```
The `@singleton` isn't necessary but is more efficient.

**Comment by jstasiak** [comment]
Yes, this is expected outcome. It may not be intuitive, but singleton is property of a binding, and you have two bindings here:

* explicit IReader -> Catalog binding
* implicit Catalog -> Catalog binding

You can work around this by either always requesting `IReader` or doing what @juharris suggests (although in that case I'd recommend to take an instance of `Catalog` as argument and return it as `IReader`, because `Catalog` is always gonna be an instance of `IReader` while not every `IReader` (in principle) has to be an instance of `Catalog`.

**Comment by Nyshiki** [comment]
Thank you very much for the quick reply.
I adopted the provider solution by @juharris and followed the suggestion of @jstasiak to invert the dependency, now the Module looks like this:
```
class CatalogModule(Module):
    def provide_ireader(self, catalog: Catalog) -> IReader:
        return catalog
```
Now both Catalog and IReader return the same instance.
I imagined the bind could have worked as an alias that linked IReader to the Catalog class, I do think the semantic of singleton is stretched a bit here, but I'm glad there's a way to do it.
Thanks again.
